RFC: NTAG I2C support

GMMan opened the pull request:
This PR is mostly to gather comments about whether the proposed architecture of NTAG I2C storage would be suitable for the project. There are gaps between some pages in NTAG I2C, so to save on space, a tag sector/page address to linear address converter is used so all the pages are packed together. Aside from that, the implementation generally works.

# What's new

- Read and emulation support for NTAG I2C 1K/2K, NTAG I2C Plus 1K/2K
- Support for NTAG I2C-specific commands
- Read/write security for NTAG I2C Plus 1K/2K, includes support for NFC_PROT, 2K_PROT, NFC_DIS_SEC1, page blanking and unauthenticated NAK on read, unauthenticated NAK on write
- SRAM support **not** implemented

# Verification 

- Read of unprotected NTAG I2C (Plus) 1K/2K tags should succeed and report the correct tag type
- Emulation of tag should result in successful read by NXP NFC TagInfo full scan
- Emulated protected tags should:
  - NAK if read starts on page specified by AUTH0
  - Regular read and fast read should have pages starting from AUTH0 blanked out with null bytes
  - PWD and PACK should be null when read out
  - Password auth should NAK if wrong password (except when tag data has zeros for PWD, continuing with current implemented behavior for other NTAGs), return PACK and allow reads/writes if successful.
  - Password protected sectors should return NAK if writing and NFC_PROT is 0
  - Entire sector 1 should NAK on read/write if NFC_DIS_SEC1 set
  - AUTLIM behavior is **not** implemented

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

GMMan committed:
nfc: Add NTAG I2C (Plus) 1K/2K read support

GMMan committed:
Fix typo

GMMan committed:
nfc: Add rudimentary NTAG I2C emulation

GMMan committed:
nfc: Closer NTAG I2C emulation plus debug logging

GMMan committed:
nfc: Fix NTAG I2C sector select emulation

GMMan committed:
nfc: Add security for NTAG I2C

GMMan committed:
nfc: Send NAK correctly for MFUL reads

GMMan committed:
nfc: Better emulate NTAG I2C SECTOR_SELECT behavior

gornekich commented:
Hello @GMMan! Thanks for your pull request!
We are waiting for i2c NTAG to be delivered to verify your feature.
I noticed that you didn't add support for saving and loading the NTAG types you introduced. Could you make it, like it's done for other types?

gornekich commented on firmware/targets/f7/furi_hal/furi_hal_nfc.c:
```diff
@@ -334,6 +335,9 @@ bool furi_hal_nfc_emulate_nfca(
                 }
                 if(buff_tx_len) {
+                    if (buff_tx_len == UINT16_MAX)
+                        buff_tx_len = 0;
```
Why do you need this?

GMMan replied:
This is to support emulating the `SECTOR_SELECT` command. After the sector number is sent in the command's second packet, the tag ACKs by not replying. It's bizarre, but that's how NXP implemented it. A normal `buff_tx_len = 0` terminates the emulation loop, so I needed a different value to indicate 0.

![image](https://user-images.githubusercontent.com/2154355/168633296-bb1f783a-711b-4efb-bcc7-0c2e3d51466e.png)

gornekich reviewed (changes requested):
Your changes break NTAG21x read functionality with message:

5824 [D][MfUltralight]: Selecting sector 255
5876 [D][FuriHalNfc]: Timeout during data exchange
5877 [D][MfUltralight]: Failed to issue sector select command
5878 [D][NfcWorker]: Failed reading Mifare Ultralight

Can you fix it?

Also, please, use `make format` to run linter

GMMan commented:
> I noticed that you didn't add support for saving and loading the NTAG types you introduced. Could you make it, like it's done for other types?

I think it's present? I added the tag types to `nfc_types.c` and currently have some saved NTAG I2Cs that I can load, view, and emulate.

GMMan committed:
nfc: Fix non-I2C Ultralight read

GMMan committed:
Merge remote-tracking branch 'origin/dev' into ntag-i2c

GMMan committed:
nfc: Reformat code

GMMan committed:
nfc: Fix non-I2C Ultralight read again

Per datasheet, max sector for SECTOR_SELECT is 0xfe, so 0xff is OK
as uninit value

gornekich committed:
Merge branch 'dev' into ntag-i2c

GMMan committed:
nfc: Only read sig for NTAG if supported

Attempting to read signature breaks immediate call to sector select on
NTAG I2C original for some reason, so don't read signature if the
command is not supported

GMMan committed:
Merge remote-tracking branch 'origin/dev' into ntag-i2c

gornekich committed:
Merge branch 'dev' into ntag-i2c
